### PR TITLE
Disallow duplicate TxOut public keys

### PR DESCRIPTION
**Pull request by eranrund**
Soundtrack of this PR: [whats currently playing in my room](https://soundcloud.com/doobdoob/doob-moloch-du-sollst-nicht-gehen-310318?in=doobdoob/sets/doob-slow)

### Motivation

We use a TxOut's public key as a unique identifier for the TxOut in numerous places, but up until now have not enforced its uniqueness. Well-behaved clients should never generate the same key twice, but it is a good idea to enforce this behavior in consensus.

### In this PR
* Adding a new TxOut-by-public-key LMDB store
* Changing validation logic across ledger db and consensus to deny having duplicate TxOut public keys.
* For the first time, adding code to gracefully migrate existing LedgerDBs so that this update is painless for live networks.

This has been deployed and slammed successfully with no noticeable performance implications. This does make the LedgerDB bigger, but thats the cost of doing business.

### CODE REVIEW

File: consensus/api/proto/consensus_common.proto [hunk 1]
```diff
@@ -56,14 +56,16 @@ enum ProposeTxResult {
...
     DuplicateKeyImages = 28;
-    MissingTxOutMembershipProof = 29;
-    InvalidTxOutMembershipProof = 30;
-    InvalidRistrettoPublicKey = 31;
-    InvalidLedgerContext = 32;
-    Ledger = 33;
-    MembershipProofValidationError = 34;
-    TxFeeError = 35;
-    KeyError = 36;
+    DuplicateOutputPublicKey = 29;
```
Comment (eranrund) [inline consensus/api/proto/consensus_common.proto:59]: This is technically a breaking change, but right now we are in control of all client implementations so this isn't gonna negatively affect anything.

File: ledger/db/src/lib.rs [hunk 2]
```diff
@@ -250,28 +278,49 @@ impl LedgerDB {
...
             .open(&path)?;
 
+        // Check if the database we opened is compatible with the current implementation.
+        let metadata_store = MetadataStore::new(&env)?;
+        let db_txn = env.begin_ro_txn()?;
+        let version = metadata_store.get_version(&db_txn)?;
+        global_log::info!("Ledger db is currently at version: {:?}", version);
+        db_txn.commit()?;
+
+        match version.is_compatible_with_latest() {
```
Comment (eranrund) [inline ledger/db/src/lib.rs:294]: This is the first time we've included such code, but I believe it is a nice experiment to have. I have verified it successfully upgrades a pre-existing LMDB database.

File: ledger/db/src/tx_out_store.rs [hunk 3]
```diff
@@ -108,7 +118,14 @@ impl TxOutStore {
             self.tx_out_index_by_hash,
             &tx_out.hash(),
             &u64_to_key_bytes(index),
-            WriteFlags::empty(),
+            WriteFlags::NO_OVERWRITE,
```
Comment (eranrund) [inline ledger/db/src/tx_out_store.rs:121]: NO_OVERWRITE ensures we do not find ourselves doing things we don't want to be doing. The `tx_out_by_*` should never have duplicate entries.

File: ledger/db/src/lib.rs (hunk 2 above)
Comment (sugargoat) [reply ledger/db/src/lib.rs:294]: Nice

File: consensus/api/proto/consensus_common.proto (hunk 1 above)
Comment (kylefleming) [reply consensus/api/proto/consensus_common.proto:59]: It's worth noting that any breaking change means server upgrades are not strictly backwards compatible w.r.t. SDKs and will require an SDK update. Is there a reason to not just add the 2 new errors at the end of the enum?

**Comment by xoloki** [comment]
If consensus will now check a local store for existing tx pubkeys, this means the consensus server knows which outputs are being created by which IP addrs.  That's bad for privacy.  But if the local store is an ORAM, that's okay.

File: consensus/api/proto/consensus_common.proto (hunk 1 above)
Comment (eranrund) [reply consensus/api/proto/consensus_common.proto:59]: Does the SDK currently match against any of these?
My preference was to use the opportunity that none of that stuff is public yet to maintain a 1:1 mapping between this and the Rust code.

**Comment by eranrund** [comment]
> If consensus will now check a local store for existing tx pubkeys, this means the consensus server knows which outputs are being created by which IP addrs. That's bad for privacy. But if the local store is an ORAM, that's okay.

I specifically asked about this in #eng yesterday and noone was concerned about it...
There are no plans to use ORAM here any time soon, or maybe at all.

File: consensus/enclave/impl/src/lib.rs [hunk 4]
```diff
@@ -422,6 +426,20 @@ impl ConsensusEnclave for SgxConsensusEnclave {
             }
         }
 
+        // Duplicate output public keys are not allowed.
+        let mut seen_output_public_keys = BTreeSet::default();
```
Comment (mfaulk) [inline consensus/enclave/impl/src/lib.rs:430]: Why is this a BTreeSet (instead of a HashSet)?

Comment (mfaulk) [reply consensus/enclave/impl/src/lib.rs:430]: I'm wondering if BTreeSet would be significantly slower than HashSet. Hard to tell:

https://doc.rust-lang.org/std/collections/struct.BTreeMap.html

"Currently, our implementation simply performs naive linear search. This provides excellent performance on small nodes of elements which are cheap to compare. However in the future we would like to further explore choosing the optimal search strategy based on the choice of B, and possibly other factors. Using linear search, searching for a random element is expected to take O(B * log(n)) comparisons, which is generally worse than a BST. In practice, however, performance is excellent."

Comment (eranrund) [reply consensus/enclave/impl/src/lib.rs:430]: Quicker lookups, similar to why `used_key_images` is also a BTreeSet. Insertion is slower, but you are doing N insertions and N^2 lookups (I think).

Comment (eranrund) [reply consensus/enclave/impl/src/lib.rs:430]: Work that was done a long time ago on optimizing performance yielded better results when using BTreeSets.

Comment (mfaulk) [reply consensus/enclave/impl/src/lib.rs:430]: Ok, that makes sense.

File: ledger/db/src/lib.rs [hunk 5]
```diff
@@ -470,6 +524,13 @@ impl LedgerDB {
             }
         }
 
+        // Check that none of the output public keys appear in the ledger.
+        for output in block_contents.outputs.iter() {
+            if self.contains_tx_out_public_key(&output.public_key)? {
```
Comment (mfaulk) [inline ledger/db/src/lib.rs:529]: There is a potential problem with this and the aggregate fee output. It is possible for an attacker to force a collision by creating an output in the ledger with a particular public key, and then forcing the enclave to use that public key as the aggregate fee's output public key. The attack is pretty contrived, but here it is:

File: ledger/db/src/lib.rs (hunk 2 above)
Comment (mfaulk) [reply ledger/db/src/lib.rs:295]: Interesting...

**Pull request merged**